File: 2020/Python/problem6.py

def part1():
    f = open("../Resources/problem6.txt")
    s = set()
    answer = 0
    for current in f.readlines():
        current = current.strip()
        if current == "\n" or current == "":
            s.clear()
        else:
            for c in current:
                if c not in s:
                    s.add(c)
                    answer += 1
    return answer
        
print(part1())

def part2():
    f = open("../Resources/problem6.txt")
    s = set()
    t = set()
    newgroup = True
    answer = 0
    for current in f.readlines():
        current = current.strip()
        if current == "\n" or current == "":
            answer += len(s)
            s.clear()
            newgroup = True
        else:
            # if this is the first line, fill the set
            if newgroup:
                newgroup = False
                for c in current:
                    s.add(c)
            else:
                for c in current:
                    t.add(c)
                s.intersection_update(t)
                t.clear()
    answer += len(s)
    return answer


print(part2()) 

File: 2020/Java/Problem6.java
import java.io.*;
import java.util.*;

class Problem6 {
    public static void main(String args[]){
        File file = new File("/home/ec2-user/environment/AOC/Resources/problem6.txt");
        
        try{
            System.out.println("Part 1: "+part1(file));
            
            System.out.println("Part 2: "+part2(file));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    public static int part1(File file) throws Exception{
        BufferedReader br = new BufferedReader(new FileReader(file));
        String current; 
        HashSet<Character> s = new HashSet<Character>();
        int answer = 0;
        while ((current = (br.readLine())) != null) {
            if(current == null || "".equals(current)){
                s.clear();
            } else {
                for(int i=0; i<current.length(); i++){
                    Character c = current.charAt(i);
                    if(!s.contains(c)){
                        s.add(c);
                        answer++;
                    }
                }
            }
        } 
        
        return answer;
    }
    
    public static int part2(File file) throws Exception{
        BufferedReader br = new BufferedReader(new FileReader(file));
        String current; 
        HashSet<Character> s = new HashSet<Character>();
        HashSet<Character> t = new HashSet<Character>();
        boolean newGroup = true;
        int answer = 0;
        while ((current = (br.readLine())) != null) {
            if(current == null || "".equals(current)){
                answer += s.size();
                s.clear();
                newGroup = true;
            } else {
                if(newGroup){
                    newGroup = false;
                    for(int i=0; i<current.length(); i++){
                        s.add(current.charAt(i));
                    }
                } else {
                    for(int i=0; i<current.length(); i++){
                        t.add(current.charAt(i));
                    }
                    s.retainAll(t);
                    t.clear();
                }
            }
        } 
        answer += s.size(); 
        
        return answer;
    }
}
